File: src/daegun/week4/베스트 앨범.swift
import Foundation

func solution(_ genres:[String], _ plays:[Int]) -> [Int] {
    
    var genreGroup: [String:[Int:Int]] = [:]
    
    // [Genre:[id:plays]]로 정리
    // genreGroup = ["classic": [3: 800, 0: 500, 2: 150], "pop": [4: 2500, 1: 600]]
    for id in 0...genres.count-1 {
        
        if genreGroup[genres[id]] == nil {
            genreGroup[genres[id]] = [id:plays[id]]
        } else {
            genreGroup[genres[id]]![id] = plays[id]
        }
    }
    
    // [Genre: 장르 총 재생횟수]로 정리
    // genreSum = ["pop": 3100, "classic": 1450]
    var genreSum: [String:Int] = [:]
    for genre in genreGroup {
        var sum = 0
        for music in genre.value {
            sum += music.value
        }
        genreSum[genre.key] = sum
    }
    
    // 장르 총 재생횟수에 따라 정렬
    // genreRank = [(key: "pop", value: 3100), (key: "classic", value: 1450)]
    let genreRank = genreSum.sorted {
        $0.1 > $1.1
    }

    // 재생 횟수가 많은 장르부터, 가장 많이 재생된 곡 2개 append
    var result: [Int] = []
    for genre in genreRank {
        
        let sortedId = genreGroup[genre.key]!.sorted { $0.0 < $1.0 }.sorted { $0.1 > $1.1 }
        if sortedId.count > 0 { result.append(sortedId[0].key) }
        if sortedId.count > 1 { result.append(sortedId[1].key) }
    }

    return result
}

File: src/jimin/week4/베스트앨범.py
from collections import defaultdict


def solution(genres, plays):
    answer = []
    album_total = defaultdict(int)
    album_list = defaultdict(list)
    # 장르별 가장 많이 재생된 음악 2개

    for idx, (g, p) in enumerate(zip(genres, plays)):
        album_total[g] += p
        album_list[g].append((p, idx))
        idx += 1
    album_total = sorted(album_total.items(), key=lambda x: x[1], reverse=True)  # 재생 횟수 정렬

    for genre, total in album_total:
        temp = sorted(album_list[genre], key=lambda x: x[0], reverse=True)[:2]
        for t in temp:
            answer.append(t[1])

    return answer


genres = ["classic", "pop", "classic", "classic", "pop"]
plays = [500, 600, 150, 800, 2500]

print(solution(genres, plays))
